chainer/testing/function_link.py

```diff
@@ -0,0 +1,1131 @@
+
+    def test_initializers(self, backend_config):
+        """Tests that the parameters of a links are correctly initialized."""
+
+        self.before_test('test_initializers')
+
+        params_inits = self._get_initializers()
+
+        # TODO(hvy): Reduce the number of loop iterations by checking
+        # multiple parameters simultaneously.
+        for i_param, param_inits in enumerate(params_inits):
+            # When testing an initializer for a particular parameter, other
+            # initializers are picked from generate_params.
+            inits = self._generate_params()
+            inits = list(inits)
+
+            for init in param_inits:
+                inits[i_param] = init
+                self._test_single_initializer(i_param, inits, backend_config)
+
+    def _get_initializers(self):
+        params_inits = self.get_initializers()
+        if not isinstance(params_inits, (tuple, list)):
+            raise TypeError(
+                '`get_initializers` must return a tuple or a list.')
+        for param_inits in params_inits:
+            if not isinstance(param_inits, (tuple, list)):
+                raise TypeError(
+                    '`get_initializers` must return a tuple or a list of '
+                    'tuples or lists.')
+            for init in param_inits:
+                _check_generated_initializer(init)
+        return params_inits
+
+    def _test_single_initializer(self, i_param, inits, backend_config):
+        # Given a set of initializer constructor arguments for the link, create
+        # and initialize a link with those arguments. `i_param` holds the index
+        # of the argument that should be tested among these.
+        inits_orig = inits
+        inits = [_get_initializer_argument_value(i) for i in inits]
+        link, _, _ = self._create_initialized_link(inits, backend_config)
+
+        # Extract the parameters from the initialized link.
+        params = _get_link_params(link, self.param_names)
+
+        # Convert the parameter of interest into a NumPy ndarray.
+        cpu_device = backend.CpuDevice()
+        param = params[i_param]
+        param_xp = param.array
+        param_np = cpu_device.send(param_xp)
+
+        # The expected values of the parameter is decided by the given
+        # initializer. If the initializer is `None`, it should have been
+        # wrapped in a InitializerArgument along with the expected initializer
+        # that the link should default to in case of `None`.
+        #
+        # Note that for this to work, the expected parameter must be inferred
+        # deterministically.
+        expected_init = _get_expected_initializer(inits_orig[i_param])
+        expected_np = numpy.empty_like(param_np)
+        expected_init(expected_np)
+
+        # Compare the values of the expected and actual parameter.
+        _check_forward_output_arrays_equal(
+            (expected_np,), (param_np,), 'forward', LinkTestError,
```

Comment:
This is not the "outputs" of the "foward". I think error message is confusing.

Reply:
How about generalizing `_check_forward_output_arrays_equal`?

Reply:
Fixed.